File: rust/src/pentomino/shape.rs
use super::{delta_swap, BitBoard, NUM_PIECES};

// Bit pattern is stored as multiple rows,
// and its line width is same as board_w.

#[derive(Clone, Debug, PartialEq)]
pub struct Shape {
    pub bitpat: BitBoard,
    pub w: usize,
    pub h: usize,
    pub ofsy: usize,
}

impl Shape {
    pub fn is_cell(&self, x: usize, y: usize, board_w: usize) -> bool {
        (self.bitpat & (1 << (y * board_w + x))) != 0
    }

    pub fn create_shapes(board_w: usize, board_h: usize) -> [(char, Vec<Self>); NUM_PIECES] {
        // F, I, L, N, P, T, U, V, W, X, Y, Z
        let original_shapes = [
            OriginalShape {
                name: 'F',
                lines: Box::new([
                    " @@",
                    "@@ ",
                    " @ ",
                ]),
            },
            OriginalShape {
                name: 'I',
                lines: Box::new([
                    "@",
                    "@",
                    "@",
                    "@",
                    "@",
                ]),
            },
            OriginalShape {
                name: 'L',
                lines: Box::new([
                    "@ ",
                    "@ ",
                    "@ ",
                    "@@",
                ]),
            },
            OriginalShape {
                name: 'N',
                lines: Box::new([
                    " @",
                    " @",
                    "@@",
                    "@ ",
                ]),
            },
            OriginalShape {
                name: 'P',
                lines: Box::new([
                    "@@",
                    "@@",
                    "@ ",
                ]),
            },
            OriginalShape {
                name: 'T',
                lines: Box::new([
                    "@@@",
                    " @ ",
                    " @ ",
                ]),
            },
            OriginalShape {
                name: 'U',
                lines: Box::new([
                    "@ @",
                    "@@@",
                ]),
            },
            OriginalShape {
                name: 'V',
                lines: Box::new([
                    "@  ",
                    "@  ",
                    "@@@",
                ]),
            },
            OriginalShape {
                name: 'W',
                lines: Box::new([
                    "@  ",
                    "@@ ",
                    " @@",
                ]),
            },
            OriginalShape {
                name: 'X',
                lines: Box::new([
                    " @ ",
                    "@@@",
                    " @ ",
                ]),
            },
            OriginalShape {
                name: 'Y',
                lines: Box::new([
                    " @",
                    "@@",
                    " @",
                    " @",
                ]),
            },
            OriginalShape {
                name: 'Z',
                lines: Box::new([
                    "@@ ",
                    " @ ",
                    " @@",
                ]),
            },
        ];

        original_shapes
            .map(|pat| (pat.name, Self::create(&pat, board_w, board_h)))
    }

    fn create(oshape: &OriginalShape, board_w: usize, board_h: usize) -> Vec<Self> {
        // Line width of bit pattern is pattern's width,
        // and convert to board_w using `to_board_width`.

        let base_shape = Self::from(&oshape.lines);

        // Create flipped and rotated shapes.
        let mut shapes = Vec::new();
        for i in 0..8 {
            let rotflipped = Self::rot_flip(&base_shape, i >= 4, i & 3);
            if rotflipped.w > board_w || rotflipped.h > board_h { continue; }
            let modified = Self {
                bitpat: Self::to_board_width(&rotflipped, board_w),
                w: rotflipped.w,
                h: rotflipped.h,
                ofsy: Self::find_ofsy(rotflipped.bitpat, rotflipped.w, rotflipped.h),
            };
            if shapes.iter().all(|s| *s != modified) {
                shapes.push(modified)
            }
        }
        shapes
    }

    fn to_board_width(shape: &Shape, board_w: usize) -> BitBoard {
        assert!(shape.w <= board_w);
        let mut bitpat = shape.bitpat;
        if shape.w < board_w {
            for y in (1..shape.h).rev() {
                let mask = (1 << (y * shape.w)) - 1;
                bitpat = (bitpat & mask) | ((bitpat & !mask) << (board_w - shape.w));
            }
        }
        bitpat
    }

    fn from(lines: &[&'static str]) -> Self {
        let w = lines[0].len();
        let h = lines.len();
        let mut bitpat = 0 as BitBoard;
        for y in 0..h {
            let bytes = lines[y].as_bytes();
            for x in 0..w {
                let c = bytes[w - x - 1];
                if c != 32 {
                    bitpat |= 1 << (y * w + x);
                }
            }
        }
        Shape { bitpat, w, h, ofsy: 0 }
    }

    fn rot_flip(base_shape: &Shape, flip: bool, rot: usize) -> Self {
        let mut shape = base_shape.clone();
        if flip {
            shape = Self::flip_y(&shape);
        }
        (0..rot).for_each(|_| {
            shape = Self::rot90(&shape);
        });
        shape
    }

    fn flip_y(shape: &Shape) -> Self {
        let mut bitpat = shape.bitpat;
        let line_mask = (1 << shape.w) - 1;
        for y in 0..shape.h / 2 {
            bitpat = delta_swap(bitpat, line_mask << (y * shape.w), (shape.h - 1 - 2 * y) * shape.w);
        }
        Shape { bitpat, w: shape.w, h: shape.h, ofsy: 0 }
    }

    fn rot90(shape: &Shape) -> Self {
        let mut bitpat = 0 as BitBoard;
        for y in 0..shape.w {
            for x in 0..shape.h {
                let b = (shape.bitpat >> ((shape.h - x - 1) * shape.w + y)) & 1;
                bitpat |= b << (y * shape.h + x);
            }
        }
        Shape { bitpat, w: shape.h, h: shape.w, ofsy: 0 }
    }

    fn find_ofsy(bitpat: BitBoard, w: usize, h: usize) -> usize {
        (0..h)
            .into_iter()
            .position(|y| (bitpat & (1 << (y * w))) != 0)
            .unwrap()
    }
}

struct OriginalShape {
    name: char,
    lines: Box<[&'static str]>,
}

File: java/src/pentomino/component/Shape.java
package pentomino.component;

import java.util.ArrayList;
import java.util.List;

import pentomino.BitBoard;

public class Shape {
    public final long bitpat;
    public final int w;
    public final int h;
    public final int ofsx;

    Shape(long bitpat, int w, int h, int ofsx) {
        this.bitpat = bitpat;
        this.w = w;
        this.h = h;
        this.ofsx = ofsx;
    }

    public String toString() {
        return "Shape(" + bitpat + ", " + w + ", " + h + ")";
    }

    public boolean equals(Object obj) {
        if (obj == null || !(obj instanceof Shape))
            return false;
        Shape other = (Shape)obj;
        return bitpat == other.bitpat && w == other.w && h == other.h;
    }

    public int hashCode() {
        return (int)((bitpat * (w * h)) % Integer.MAX_VALUE);
    }

    //

    static Shape[] createFromOriginal(OriginalShape os, int boardW, int boardH) {
        Shape baseShape = createBase(os, boardW, boardH);

        List<Shape> shapes = new ArrayList<Shape>();
        for (int i = 0; i < 8; ++i) {
                Shape rotflipped = rotFlip(baseShape, i >= 4, i & 3);
            if (rotflipped.w > boardW || rotflipped.h > boardH)
                continue;

                Shape modified = new Shape(
                toBoardWidth(rotflipped, boardW),
                rotflipped.w,
                rotflipped.h,
                findOfsx(rotflipped.bitpat, rotflipped.w, rotflipped.h)
            );
            if (!shapes.contains(modified))
                shapes.add(modified);
        }
        return shapes.toArray(Shape[]::new);
    }

    static long toBoardWidth(Shape shape, int board_w) {
        // assert!(shape.w <= board_w);
        long bitpat = shape.bitpat;
        if (shape.w < board_w) {
            for (int y = shape.h; --y > 0; ) {
                long mask = (1L << (y * shape.w)) - 1;
                bitpat = (bitpat & mask) | ((bitpat & ~mask) << (board_w - shape.w));
            }
        }
        return bitpat;
    }

    static Shape rotFlip(Shape baseShape, boolean flip, int rot) {
        Shape shape = baseShape;
        if (flip)
            shape = flipY(shape);
        for (int i = 0; i < rot; ++i)
            shape = rot90(shape);
        return shape;
    }

    static Shape flipY(Shape shape) {
        long bitpat = shape.bitpat;
        long line_mask = (1 << shape.w) - 1;
        for (int y = 0; y < shape.h / 2; ++y) {
            bitpat = BitBoard.deltaSwap(bitpat, line_mask << (y * shape.w), (shape.h - 1 - 2 * y) * shape.w);
        }
        return new Shape(bitpat, shape.w, shape.h, 0);
    }

    static Shape rot90(Shape shape) {
        long bitpat = 0;
        for (int y = 0; y < shape.w; ++y) {
            for (int x = 0; x < shape.h; ++x) {
                long b = (shape.bitpat >> ((shape.h - x - 1) * shape.w + y)) & 1;
                bitpat |= b << (y * shape.h + x);
            }
        }
        return new Shape(bitpat, shape.h, shape.w, 0);
    }

    static int findOfsx(long bitpat, int w, int _h) throws AssertionError {
        for (int i = 0; i < w; ++i) {
            if ((bitpat & (1L << i)) != 0)
                return i;
        }
        assert false;
        return -1;
    }

    static Shape createBase(OriginalShape os, int boardW, int boardH) {
        int w = os.lines[0].length();
        int h = os.lines.length;
        long bitpat = 0;
        for (int y = 0; y < h; ++y) {
            String bytes = os.lines[y];
            for (int x = 0; x < w; ++x) {
                long c = bytes.charAt(w - x - 1);
                if (c != 32) {
                    bitpat |= 1 << (y * w + x);
                }
            }
        }
        return new Shape(bitpat, w, h, 0);
    }
}
